src/snowflake/cli/plugins/nativeapp/artifacts.py

```diff
@@ -75,37 +60,300 @@ class NotInDeployRootError(ClickException):
+            return
+
+        canonical_src = self._canonical_src(src)
+        canonical_dest = self._canonical_dest(dest)
+
+        src_is_dir = absolute_src.is_dir()
+        if map_as_child:
+            # Make sure the destination is a child of the original, since this was requested
+            canonical_dest = canonical_dest / canonical_src.name
+            dest_is_dir = src.is_dir()
+
+        if canonical_src in self._src_to_dest:
+            raise TooManyFilesError(canonical_dest)
+
+        # Verify that multiple files are not being mapped to a single file destination
+        current_sources = self._dest_to_src.setdefault(canonical_dest, [])
+        if not dest_is_dir:
+            # the destination is a file
+            if (canonical_src not in current_sources) and len(current_sources) > 0:
+                raise TooManyFilesError(dest)
+
+        # Perform all updates together we don't end up with inconsistent state
+        self._update_dest_is_dir(canonical_dest, dest_is_dir)
+        self._src_to_dest[canonical_src] = canonical_dest
+        if canonical_src not in current_sources:
+            current_sources.append(canonical_src)
+
+    def _add_mapping(self, src: str, dest: Optional[str]):
+        """
+        Adds the specified artifact rule to this instance. The source should be relative to the project directory. It
+        is interpreted as a file, directory or glob pattern. If the destination path is not specified, each source match
+        is mapped to an identical path in the deploy root.
+        """
+        match_found = False
+
+        src_path = Path(src)
+        if src_path.is_absolute():
+            raise ArtifactError("Source path must be a relative path")
+
+        for resolved_src in self._project_root.glob(src):
+            match_found = True
+
+            if dest:
+                dest_stem = dest.rstrip("/")
+                if not dest_stem:
+                    # handle '/' as the destination as a special case
+                    raise NotInDeployRootError(dest, self._deploy_root)
+                dest_path = Path(dest.rstrip("/"))
+                if dest_path.is_absolute():
+                    raise ArtifactError("Destination path must be a relative path")
+                self._add(resolved_src, dest_path, specifies_directory(dest))
+            else:
+                self._add(
+                    resolved_src,
+                    resolved_src.relative_to(self._project_root),
+                    False,
+                )
+
+        if not match_found:
+            raise SourceNotFoundError(src)
+
+    def add(self, mapping: Union[ArtifactMapping, PathMapping]) -> None:
+        """
+        Adds an artifact mapping rule to this instance.
+        """
+        if isinstance(mapping, ArtifactMapping):
+            self._add_mapping(mapping.src, mapping.dest)
+        elif isinstance(mapping, PathMapping):
+            self._add_mapping(mapping.src, mapping.dest)
```

Comment:
Not clear on why the distinction here. I thought part of `PathMapping` is used to create `ArtifactMapping` ?

Reply:
Think of it as 2 distinct APIs for now. Our existing code uses ArtifactMapping everywhere, but the upcoming use by processors will rely on PathMapping instead. So I handle both, until I can unify them.